integrations/airflow/marquez_airflow/extractors/great_expectations_extractor.py

```diff
@@ -0,0 +1,347 @@
+            # Great expectations generate long, impenetrable name here
+            validation_result = next(iter(run_result.values()))['validation_result']
+
+            data_quality_facet = self.parse_data_quality_facet(validation_result)
+            if not data_quality_facet:
+                return None
+
+            batch_kwargs = get_from_nullable_chain(
+                validation_result,
+                ['meta', 'batch_kwargs']
+            )
+            # To match dataset name we need canonical datasource name
+            name = batch_kwargs.get('datasource', None)
+            path = batch_kwargs.get('path', None)
+            return Dataset(
+                source=Source(
+                    name=name,
+                    type='GREAT_EXPECTATIONS',
+                    connection_url=path
+                ),
+                name=name,
+                type=DatasetType.DB_TABLE,
+                custom_facets={
+                    'dataQuality': data_quality_facet
+                }
+            )
+
+        except ValueError:
+            pass
+        return None
+
+    def parse_data_quality_facet(self, validation_result: Dict) \
+            -> Optional[DataQualityDatasetFacet]:
+        facet_data = {
+            "columnMetrics": defaultdict(dict)
+        }
+
+        # try to get to actual expectations results
+        try:
+            expectations_results = validation_result['results']
+            for expectation in expectations_results:
+                for parser in _EXPECTATIONS_PARSERS:
+
+                    # accept possible duplication, should have no difference in results
+                    if parser.can_accept(expectation):
+                        result = parser.parse_expectation_result(expectation)
+                        facet_data[result.facet_key] = result.value
+                for parser in _COLUMN_EXPECTATIONS_PARSER:
+                    if parser.can_accept(expectation):
+                        result = parser.parse_expectation_result(expectation)
+                        facet_data['columnMetrics'][result.column_id][result.facet_key] \
+                            = result.value
+
+            for key in facet_data['columnMetrics'].keys():
+                facet_data['columnMetrics'][key] = ColumnMetric(**facet_data['columnMetrics'][key])
+            return DataQualityDatasetFacet(**facet_data)
+        except ValueError:
+            log.exception(
+                "Great Expectations's CheckpointResult object does not have expected key"
+            )
+        return None
+
+    def extract(self) -> Optional[StepMetadata]:
+        return None
+
+    def extract_on_complete(self, task_instance) -> Optional[StepMetadata]:
+        if self.result:
+            dataset = self.parse_result()
+            if not dataset:
+                return None
+            return StepMetadata(
+                name=get_job_name(task=self.operator),
+                outputs=[dataset]
+            )
```

Comment:
I think this should be inputs, not outputs. The Great Expectations job doesn't create the dataset- it reads it. Marking the dataset as an output will create a new dataset version, which may or may not share lineage with other downstream consumers (depending on which job completes first).

Reply:
yeah, we can search for input facets for a given dataset version, so there shouldn't be any concern there.